presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/OrcStorageManager.java

```diff
@@ -0,0 +1,594 @@
+            implements StoragePageSink
+    {
+        private final long transactionId;
+        private final long tableId;
+        private final int bucketNumber;
+        private final List<Long> columnIds;
+        private final List<Type> columnTypes;
+        private final CompressionType compressionType;
+
+        private final List<File> stagingFiles = new ArrayList<>();
+        private final List<ChunkInfo> chunks = new ArrayList<>();
+        private final List<ListenableFuture<?>> chunkStoreJobs = new ArrayList<>();
+
+        private boolean committed;
+        private OrcFileWriter writer;
+        private long chunkId;
+
+        public OrcStoragePageSink(
+                long transactionId,
+                long tableId,
+                int bucketNumber,
+                List<Long> columnIds,
+                List<Type> columnTypes,
+                CompressionType compressionType)
+        {
+            this.transactionId = transactionId;
+            this.tableId = tableId;
+            this.bucketNumber = bucketNumber;
+            this.columnIds = ImmutableList.copyOf(requireNonNull(columnIds, "columnIds is null"));
+            this.columnTypes = ImmutableList.copyOf(requireNonNull(columnTypes, "columnTypes is null"));
+            this.compressionType = requireNonNull(compressionType, "compressionType is null");
+        }
+
+        @Override
+        public void appendPages(List<Page> pages)
+        {
+            createWriterIfNecessary();
+            writer.appendPages(pages);
+        }
+
+        @Override
+        public void appendPages(List<Page> pages, int[] pageIndexes, int[] positionIndexes)
+        {
+            createWriterIfNecessary();
+            writer.appendPages(pages, pageIndexes, positionIndexes);
+        }
+
+        @Override
+        public boolean isFull()
+        {
+            if (writer == null) {
+                return false;
+            }
+            return (writer.getRowCount() >= maxChunkRows) || (writer.getUncompressedSize() >= maxChunkSize.toBytes());
+        }
+
+        @Override
+        public void flush()
+        {
+            if (writer != null) {
+                writer.close();
+
+                File file = storageService.getStagingFile(chunkId);
+
+                chunkRecorder.recordCreatedChunk(transactionId, tableId, chunkId, file.length());
+
+                chunkStoreJobs.add(chunkStoreManager.submit(chunkId, file));
+
+                long rowCount = writer.getRowCount();
+                long uncompressedSize = writer.getUncompressedSize();
+
+                chunks.add(createChunkInfo(chunkId, bucketNumber, file, rowCount, uncompressedSize));
+
+                writer = null;
+                chunkId = 0;
+            }
+        }
+
+        @Override
+        public ListenableFuture<List<ChunkInfo>> commit()
+        {
+            checkState(!committed, "already committed");
+            committed = true;
+
+            flush();
```

Comment:
What if it fails? Do we wanna call `rollback` like what we are doing in some Hive writers?